File: JavaAPI/v3.14.2.1/examples/BPipe/com/bloomberglp/blpapi/examples/SimpleCategorizedFieldSearchExample.java
/* Copyright 2012. Bloomberg Finance L.P.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:  The above copyright notice and this
 * permission notice shall be included in all copies or substantial portions of
 * the Software.  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
 * EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
 * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */
package com.bloomberglp.blpapi.examples;

import java.util.ArrayList;
import java.io.IOException;

import com.bloomberglp.blpapi.Event;
import com.bloomberglp.blpapi.Event.EventType;
import com.bloomberglp.blpapi.EventQueue;
import com.bloomberglp.blpapi.CorrelationID;
import com.bloomberglp.blpapi.Message;
import com.bloomberglp.blpapi.MessageIterator;
import com.bloomberglp.blpapi.Request;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Service;
import com.bloomberglp.blpapi.Session;
import com.bloomberglp.blpapi.SessionOptions;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Identity;
import com.bloomberglp.blpapi.SessionOptions.ServerAddress;

public class SimpleCategorizedFieldSearchExample {

    private static final String APIFLDS_SVC = "//blp/apiflds";
	private static final String AUTH_SVC = "//blp/apiauth";

    private static final int CAT_NAME_LEN   = 40;
    private static final int ID_LEN         = 13;
    private static final int MNEMONIC_LEN   = 36;
    private static final int DESC_LEN       = 40;
    private static final String PADDING     = 
        "                                            ";
    private static final Name FIELD_ID       = new Name("id");
    private static final Name FIELD_MNEMONIC = new Name("mnemonic");
    private static final Name FIELD_DATA     = new Name("fieldData");
    private static final Name FIELD_DESC     = new Name("description");
    private static final Name FIELD_INFO     = new Name("fieldInfo");
    private static final Name FIELD_ERROR    = new Name("fieldError");
    private static final Name FIELD_MSG      = new Name("message");
    private static final Name CATEGORY       = new Name("category");
    private static final Name CATEGORY_NAME  = new Name("categoryName");
    private static final Name CATEGORY_ID    = new Name("categoryId");
    private static final Name FIELD_SEARCH_ERROR = new Name("fieldSearchError");
	private static final Name AUTHORIZATION_SUCCESS = Name.getName("AuthorizationSuccess");
	private static final Name TOKEN_SUCCESS = Name.getName("TokenGenerationSuccess");

	private ArrayList<String>       d_hosts;
    private int                     d_port;
    private String                  d_authOption;
    private String                  d_name;
    private Identity                d_identity;
    private Session           		d_session;

    private boolean createSession()	throws IOException, InterruptedException 
    {
		String authOptions = null;
		if(d_authOption.equalsIgnoreCase("APPLICATION")){
	        // Set Application Authentication Option
	        authOptions = "AuthenticationMode=APPLICATION_ONLY;";
	        authOptions += "ApplicationAuthenticationType=APPNAME_AND_KEY;";
	        // ApplicationName is the entry in EMRS.
	        authOptions += "ApplicationName=" + d_name;
		} else {
	        // Set User authentication option
	        if (d_authOption.equalsIgnoreCase("NONE"))
	        {
	        	d_authOption = null;
	        }
	        else
	        {
	            if (d_authOption.equalsIgnoreCase("USER_APP"))
	            {
	                // Set User and Application Authentication Option
	                authOptions = "AuthenticationMode=USER_AND_APPLICATION;";
	                authOptions += "AuthenticationType=OS_LOGON;";
	                authOptions += "ApplicationAuthenticationType=APPNAME_AND_KEY;";
	                // ApplicationName is the entry in EMRS.
	                authOptions += "ApplicationName=" + d_name;
	            }
	            else
	            {
		            if (d_authOption.equalsIgnoreCase("DIRSVC"))
		            {
		                // Authenticate user using active directory service property
		                authOptions = "AuthenticationType=DIRECTORY_SERVICE;";
		                authOptions += "DirSvcPropertyName=" + d_name;
		            }
		            else
		            {
		                // Authenticate user using windows/unix login name
		                authOptions = "AuthenticationType=OS_LOGON";
		            }
	            }
	        }
		}
	
		System.out.println("authOptions = " + authOptions);
		SessionOptions sessionOptions = new SessionOptions();
		if (d_authOption != null)
		{
			sessionOptions.setAuthenticationOptions(authOptions);
		}
	
		ServerAddress[] servers = new ServerAddress[d_hosts.size()];
		for (int i = 0; i < d_hosts.size(); ++i) {
			servers[i] = new ServerAddress(d_hosts.get(i), d_port);
		}
	
		sessionOptions.setServerAddresses(servers);
	    sessionOptions.setAutoRestartOnDisconnection(true);
	    sessionOptions.setNumStartAttempts(d_hosts.size());
		
		System.out.print("Connecting to port " + d_port + " on server:");
		for (ServerAddress server: sessionOptions.getServerAddresses()) {
			System.out.print(" " + server);
		}
		System.out.println();
	    d_session = new Session(sessionOptions);
	    
	    return d_session.start();
	}
	
	private boolean authorize()
		throws IOException, InterruptedException {
		Event event;
		MessageIterator msgIter;
	
		EventQueue tokenEventQueue = new EventQueue();
		CorrelationID corrlationId = new CorrelationID(99);
		d_session.generateToken(corrlationId, tokenEventQueue);
		String token = null;
		int timeoutMilliSeonds = 10000;
		event = tokenEventQueue.nextEvent(timeoutMilliSeonds);
		if (event.eventType() == EventType.TOKEN_STATUS) {
			MessageIterator iter = event.messageIterator();
			while (iter.hasNext()) {
				Message msg = iter.next();
				System.out.println(msg.toString());
				if (msg.messageType() == TOKEN_SUCCESS) {
					token = msg.getElementAsString("token");
				}
			}
		}
		if (token == null){
			System.err.println("Failed to get token");
			return false;
		}
	
		if (d_session.openService(AUTH_SVC)) {
			Service authService = d_session.getService(AUTH_SVC);
			Request authRequest = authService.createAuthorizationRequest();
			authRequest.set("token", token);
	
			EventQueue authEventQueue = new EventQueue();
	
			d_session.sendAuthorizationRequest(authRequest, d_identity,
					authEventQueue, new CorrelationID(d_identity));
	
			while (true) {
				event = authEventQueue.nextEvent();
				if (event.eventType() == EventType.RESPONSE
						|| event.eventType() == EventType.PARTIAL_RESPONSE
						|| event.eventType() == EventType.REQUEST_STATUS) {
					msgIter = event.messageIterator();
					while (msgIter.hasNext()) {
						Message msg = msgIter.next();
						System.out.println(msg);
						if (msg.messageType() == AUTHORIZATION_SUCCESS) {
							return true;
						} else {
							System.err.println("Not authorized");
							return false;
						}
					}
				}
			}
		}
		return false;
	}

    public static void main(String[] args) throws Exception
    {
        SimpleCategorizedFieldSearchExample example = 
            new SimpleCategorizedFieldSearchExample();
        example.run(args);
        
        System.out.println("Press ENTER to quit");
        System.in.read();
    }

    private void run(String[] args) throws Exception
    {
    	d_hosts = new ArrayList<String>();
        d_port = 8194;
        d_authOption="";
		d_name="";
		d_session = null;

        if (!parseCommandLine(args)) {
            return ;
        }

        if (!createSession()) {
            System.err.println("Failed to start session.");
            return;
        }

		if (d_authOption != null) {
			d_identity = d_session.createIdentity();
			if (!authorize()) {
				return;
			}
		}

        System.out.println("Connected successfully.");
        if (!d_session.openService(APIFLDS_SVC)) {
        	System.err.println("Failed to open " + APIFLDS_SVC);
        	d_session.stop();
        	return;
        }

        Service fieldInfoService = d_session.getService(APIFLDS_SVC);
        Request request = fieldInfoService.createRequest(
        "CategorizedFieldSearchRequest");
        request.set ("searchSpec", "last price");
        Element exclude = request.getElement("exclude");
        exclude.setElement ("fieldType", "Static");
        request.set ("returnFieldDocumentation", false);

		if (d_authOption == null)
		{
            System.out.println("Sending Request: " + request);
			d_session.sendRequest(request, null);
		}
		else
		{
			// request data with identity object
            System.out.println("Sending Request with user's Identity: " + request);
			d_session.sendRequest(request, d_identity, null);
		}

        boolean bGotResponse = false;
        while (!bGotResponse) {
            try {
                Event event = d_session.nextEvent();
                MessageIterator msgIter = event.messageIterator();

                while (msgIter.hasNext()) {
                    Message msg = msgIter.next();

                    if (event.eventType() != Event.EventType.RESPONSE &&
                            event.eventType() != Event.EventType.PARTIAL_RESPONSE) {
                        continue;
                    }
                    if (msg.hasElement(FIELD_SEARCH_ERROR)) {
                        System.out.println(msg);
                        continue;
                    }
                    Element categories = msg.getElement(CATEGORY);
                    int numCategories = categories.numValues();

                    for (int catIdx=0; catIdx < numCategories; ++catIdx) {

                        Element category = categories.getValueAsElement(catIdx);
                        String Name = category.getElementAsString(
                                CATEGORY_NAME);
                        String Id = category.getElementAsString(CATEGORY_ID);

                        System.out.println ("\n  Category Name:" +
                                padString (Name, CAT_NAME_LEN) +
                                "\tId:" + Id) ;

                        Element fields = category.getElement(FIELD_DATA);
                        int numElements = fields.numValues();

                        printHeader();
                        for (int i=0; i < numElements; i++) {
                            printField (fields.getValueAsElement(i));
                        }
                    }
                    System.out.println();
                }
                if (event.eventType() == Event.EventType.RESPONSE) break;
            } 

            catch (Exception ex) {
                System.out.println ("Got Exception:" + ex);
            }
        }
    }

    private void printField (Element field)
    {
        String fldId, fldMnemonic, fldDesc;

        fldId       = field.getElementAsString(FIELD_ID);
        if (field.hasElement(FIELD_INFO)) {
            Element fldInfo     = field.getElement (FIELD_INFO) ;
            fldMnemonic = fldInfo.getElementAsString(FIELD_MNEMONIC);
            fldDesc     = fldInfo.getElementAsString(FIELD_DESC);

            System.out.println( padString(fldId, ID_LEN) +
                    padString (fldMnemonic, MNEMONIC_LEN) +
                    padString (fldDesc, DESC_LEN));
        }
        else {
            Element fldError = field.getElement(FIELD_ERROR) ;
            fldDesc = fldError.getElementAsString(FIELD_MSG) ;

            System.out.println ("\n ERROR: " + fldId + " - " + fldDesc) ;
        }
    }

    private void printHeader ()
    {
        System.out.println( padString("FIELD ID", ID_LEN) +
                padString("MNEMONIC", MNEMONIC_LEN) +
                padString("DESCRIPTION", DESC_LEN));
        System.out.println( padString("-----------", ID_LEN) +
                padString("-----------", MNEMONIC_LEN) +
                padString("-----------", DESC_LEN));
    }

    private static String padString(String str, int width)
    {
        if (str.length() >= width || str.length() >= PADDING.length() ) return str;
        else return str + PADDING.substring(0, width-str.length());
    }
    private boolean parseCommandLine(String[] args)
    {
        for (int i = 0; i < args.length; ++i)
        {
        	if (args[i].equalsIgnoreCase("-ip") && i + 1 < args.length)
            {
                d_hosts.add(args[++i]);
            }
            else if (args[i].equalsIgnoreCase("-p") && i + 1 < args.length)
            {
                d_port = Integer.parseInt(args[++i]);
            }
			else if(args[i].equalsIgnoreCase("-auth") && i + 1 < args.length) {
				d_authOption = args[++i];
			}
			else if(args[i].equalsIgnoreCase("-n") && i + 1 < args.length) {
				d_name = args[++i];
			}
            else if (args[i].equalsIgnoreCase("-h"))
            {
                printUsage();
				return false;
            }
        }

        // check for host ip
        if (d_hosts.size() == 0)
        {
        	System.out.println("Missing host IP");
        	printUsage();
        	return false;
        }

        // check for application name
        if ((d_authOption.equalsIgnoreCase("APPLICATION")  || d_authOption.equalsIgnoreCase("USER_APP")) && (d_name.equalsIgnoreCase("")))
        {
        	System.out.println("Application name cannot be NULL for application authorization.");
            printUsage();
            return false;
        }
        // check for Directory Service name
        if ((d_authOption.equalsIgnoreCase("DIRSVC")) && (d_name.equalsIgnoreCase("")))
        {
        	System.out.println("Directory Service property name cannot be NULL for DIRSVC authorization.");
            printUsage();
            return false;
        }
        return(true);
    }

    private void printUsage()
    {
        System.out.println("Usage:");
        System.out.println(
        	"	Retrieve field information in categorized form ");
        System.out.println("        [-ip        <ipAddress	= localhost>");
        System.out.println("        [-p         <tcpPort	= 8194>");
		System.out.println("        [-auth      <authenticationOption = LOGON (default) or NONE or APPLICATION or DIRSVC or USER_APP>]");
		System.out.println("        [-n         <name = applicationName or directoryService>]");
		System.out.println("Notes:");
		System.out.println(" -Specify only LOGON to authorize 'user' using Windows login name.");
		System.out.println(" -Specify DIRSVC and name(Directory Service Property) to authorize user using directory Service.");
		System.out.println(" -Specify APPLICATION and name(Application Name) to authorize application.");
    }
}

File: C++API/v3.14.3.1/examples/BPipe/SimpleCategorizedFieldSearchExample.cpp
/* Copyright 2012. Bloomberg Finance L.P.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:  The above copyright notice and this
 * permission notice shall be included in all copies or substantial portions of
 * the Software.  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
 * EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
 * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */
#include <blpapi_session.h>
#include <blpapi_eventdispatcher.h>

#include <blpapi_event.h>
#include <blpapi_message.h>
#include <blpapi_element.h>
#include <blpapi_name.h>
#include <blpapi_request.h>
#include <blpapi_subscriptionlist.h>
#include <blpapi_defs.h>
#include <blpapi_exception.h>

#include <iostream>
#include <vector>
#include <string>
#include <stdlib.h>
#include <string.h>

using namespace BloombergLP;
using namespace blpapi;

namespace {
    const Name AUTHORIZATION_SUCCESS("AuthorizationSuccess");
    const Name AUTHORIZATION_FAILURE("AuthorizationFailure");
    const Name TOKEN_SUCCESS("TokenGenerationSuccess");
    const Name TOKEN_FAILURE("TokenGenerationFailure");
	const Name TOKEN("token");

	const Name FIELD_ID("id");
    const Name FIELD_MNEMONIC("mnemonic");
    const Name FIELD_DATA("fieldData");
    const Name FIELD_DESC("description");
    const Name FIELD_INFO("fieldInfo");
    const Name FIELD_ERROR("fieldError");
    const Name FIELD_MSG("message");
    const Name CATEGORY("category");
    const Name CATEGORY_NAME("categoryName");
    const Name CATEGORY_ID("categoryId");
    const Name FIELD_SEARCH_ERROR("fieldSearchError");

	const char *AUTH_SVC = "//blp/apiauth";
};

class SimpleCategorizedFieldSearchExample
{
    int ID_LEN;
    int MNEMONIC_LEN;
    int DESC_LEN;
    int CAT_NAME_LEN;
    std::string PADDING;
    std::string APIFLDS_SVC;
	std::vector<std::string> d_hosts;		// IP Addresses of appliances
    int					d_port;
	std::string			d_authOption;	// authentication option user/application
	std::string			d_name;	        // DirectoryService/ApplicationName
    Identity			d_identity;
	Session				*d_session;

    void printUsage()
    {
        std::cout << "Usage:" << std::endl
            << "    Categorized Field Search Example " << std::endl
            << "        [-ip        <ipAddress  = localhost>" << std::endl
            << "        [-p         <tcpPort    = 8194>" << std::endl
			<< "        [-auth      <authenticationOption = LOGON (default) or NONE or APPLICATION or DIRSVC or USER_APP>]" << std::endl
            << "        [-n         <name = applicationName or directoryService>]" << std::endl
            << "Notes:" << std::endl
            << " -Specify only LOGON to authorize 'user' using Windows login name." << std::endl
            << " -Specify DIRSVC and name(Directory Service Property) to authorize user using directory Service." << std::endl
            << " -Specify APPLICATION and name(Application Name) to authorize application." << std::endl;
    }

    bool parseCommandLine(int argc, char **argv)
    {
        for (int i = 1; i < argc; ++i) {
            if (!std::strcmp(argv[i],"-ip") && i + 1 < argc) {
                d_hosts.push_back(argv[++i]);
            }else if (!std::strcmp(argv[i],"-p") &&  i + 1 < argc) {
                d_port = std::atoi(argv[++i]);
			} else if (!std::strcmp(argv[i],"-auth") &&  i + 1 < argc) {
				d_authOption = argv[++i];
			} else if (!std::strcmp(argv[i],"-n") &&  i + 1 < argc) {
				d_name = argv[++i];
			} else if (!std::strcmp(argv[i], "-h")) {
				printUsage();
				return false;
			}
        }
		// check for hosts
        if (d_hosts.size() == 0) {
			 std::cout << "Missing host ip." << std::endl;
			 printUsage();
             return false;
        }

		// check for appliation name
		if ((!std::strcmp(d_authOption.c_str(),"APPLICATION") || !std::strcmp(d_authOption.c_str(), "USER_APP")) && (!std::strcmp(d_name.c_str(), ""))){
			 std::cout << "Application name cannot be NULL for application authorization." << std::endl;
			 printUsage();
             return false;
		}
		// check for Directory Service name
		if ((!std::strcmp(d_authOption.c_str(),"DIRSVC")) && (!std::strcmp(d_name.c_str(), ""))){
			 std::cout << "Directory Service property name cannot be NULL for DIRSVC authorization." << std::endl;
			 printUsage();
             return false;
		}
        return true;
    }

    std::string padString(std::string str, unsigned int width)
    {
        if (str.length() >= width || str.length() >= PADDING.length() ) return str;
        else return str + PADDING.substr(0, width-str.length());
    }

    void printField (const Element &field)
    {
        std::string  fldId       = field.getElementAsString(FIELD_ID);
        if (field.hasElement(FIELD_INFO)) {
            Element fldInfo          = field.getElement (FIELD_INFO) ;
            std::string  fldMnemonic = 
                fldInfo.getElementAsString(FIELD_MNEMONIC);
            std::string  fldDesc     =
                fldInfo.getElementAsString(FIELD_DESC);

            std::cout << padString(fldId, ID_LEN) 
                << padString (fldMnemonic, MNEMONIC_LEN)
                << padString (fldDesc, DESC_LEN) << std::endl;
        }
        else {
            Element fldError = field.getElement(FIELD_ERROR) ;
            std::string  errorMsg = fldError.getElementAsString(FIELD_MSG) ;

            std::cout << std::endl << " ERROR: " << fldId << " - "
                << errorMsg << std::endl ;
        }
    }

    void printHeader ()
    {
        std::cout << padString("FIELD ID", ID_LEN) +
            padString("MNEMONIC", MNEMONIC_LEN) +
            padString("DESCRIPTION", DESC_LEN)
            << std::endl;
        std::cout << padString("-----------", ID_LEN) +
            padString("-----------", MNEMONIC_LEN) +
            padString("-----------", DESC_LEN)
            << std::endl;
    }

	bool startSession(){

        SessionOptions sessionOptions;
		std::string authOptions;

		std::cout << "Connecting to port " << d_port << " on server: "; 
        for (size_t i = 0; i < d_hosts.size(); ++i) {
            sessionOptions.setServerAddress(d_hosts[i].c_str(), d_port, i);
			std::cout << d_hosts[i].c_str();
        }
		std::cout << std::endl;

        sessionOptions.setServerPort(d_port);
        sessionOptions.setAutoRestartOnDisconnection(true);
        sessionOptions.setNumStartAttempts(d_hosts.size());
		authOptions = getAuthOptions();
		if (authOptions.size() > 0) {
			sessionOptions.setAuthenticationOptions(authOptions.c_str());
		}
        d_session = new Session(sessionOptions);

        if (!d_session->start()) {
            std::cerr << "Failed to connect!" << std::endl;
            return false;
        }
        
        return true;
    }

	// construct authentication option string
	std::string getAuthOptions()
	{
		std::string authOptions;
		if (!std::strcmp(d_authOption.c_str(),"APPLICATION")) { //  Authenticate application
			// Set Application Authentication Option
			authOptions = "AuthenticationMode=APPLICATION_ONLY;";
			authOptions+= "ApplicationAuthenticationType=APPNAME_AND_KEY;";
			// ApplicationName is the entry in EMRS.
			authOptions+= "ApplicationName=" + d_name;
		} else if (!strcmp(d_authOption.c_str(), "NONE")) {
			// do nothing
		} else if (!strcmp(d_authOption.c_str(), "USER_APP")) {
            // Set User and Application Authentication Option
            authOptions = "AuthenticationMode=USER_AND_APPLICATION;";
            authOptions += "AuthenticationType=OS_LOGON;";
            authOptions += "ApplicationAuthenticationType=APPNAME_AND_KEY;";
            // ApplicationName is the entry in EMRS.
            authOptions += "ApplicationName=" + d_name;
		} else if (!strcmp(d_authOption.c_str(), "DIRSVC")) {		
				// Authenticate user using active directory service property
				authOptions = "AuthenticationType=DIRECTORY_SERVICE;";
				authOptions += "DirSvcPropertyName=" + d_name;
		} else {
				// Authenticate user using windows/unix login name
				authOptions = "AuthenticationType=OS_LOGON";
		}

		std::cout << "Authentication Options = " << authOptions << std::endl;
		return authOptions;
	}

	bool authorize()
    {
        EventQueue tokenEventQueue;
        d_session->generateToken(CorrelationId(), &tokenEventQueue);
        std::string token;
        Event event = tokenEventQueue.nextEvent();
        if (event.eventType() == Event::TOKEN_STATUS) {
            MessageIterator iter(event);
            while (iter.next()) {
                Message msg = iter.message();
                msg.print(std::cout);
                if (msg.messageType() == TOKEN_SUCCESS) {
                    token = msg.getElementAsString(TOKEN);
                }
                else if (msg.messageType() == TOKEN_FAILURE) {
                    break;
                }
            }
        }
        if (token.length() == 0) {
            std::cout << "Failed to get token" << std::endl;
            return false;
        }

        d_session->openService(AUTH_SVC);
		Service authService = d_session->getService(AUTH_SVC);
        Request authRequest = authService.createAuthorizationRequest();
        authRequest.set(TOKEN, token.c_str());

        EventQueue authQueue;
		d_identity = d_session->createIdentity();
        d_session->sendAuthorizationRequest(
            authRequest, &d_identity, CorrelationId(), &authQueue);

        while (true) {
            Event event = authQueue.nextEvent();
            if (event.eventType() == Event::RESPONSE ||
                event.eventType() == Event::REQUEST_STATUS ||
                event.eventType() == Event::PARTIAL_RESPONSE) {
                    MessageIterator msgIter(event);
                    while (msgIter.next()) {
                        Message msg = msgIter.message();
                        msg.print(std::cout);
                        if (msg.messageType() == AUTHORIZATION_SUCCESS) {
                            return true;
                        }
                        else {
                            std::cout << "Authorization failed" << std::endl;
                            return false;
                        }
                    }
            }
        }
    }

public:
      SimpleCategorizedFieldSearchExample(): d_session(0)
		  , PADDING("                                            ")
          , APIFLDS_SVC("//blp/apiflds")
		  , d_name("") 
	  {
              ID_LEN         = 13;
              MNEMONIC_LEN   = 36;
              DESC_LEN       = 40;
              CAT_NAME_LEN   = 40;
      }
	  
	  ~SimpleCategorizedFieldSearchExample()
	  {
		  if (d_session) delete d_session;
	  }

      void run(int argc, char **argv)
      {
          d_port = 8194;
          if (!parseCommandLine(argc, argv)) return;

		  if (!startSession()) return;

		  if (strcmp(d_authOption.c_str(), "NONE")) {
			  if (!authorize()) {
				return;
			  }
		  }
          if (!d_session->openService(APIFLDS_SVC.c_str())) {
              std::cerr <<"Failed to open " << APIFLDS_SVC << std::endl;
              return;
          }

          Service fieldInfoService = d_session->getService(APIFLDS_SVC.c_str());
          Request request = fieldInfoService.createRequest(
              "CategorizedFieldSearchRequest");
          request.set ("searchSpec", "last price");
          Element exclude = request.getElement("exclude");
          exclude.setElement("fieldType", "Static");
          request.set ("returnFieldDocumentation", false);

		  if (!strcmp(d_authOption.c_str(), "NONE")) {
			  std::cout << "Sending Request: "  << request << std::endl;
			  d_session->sendRequest(request, CorrelationId(this));
		} else {
			std::cout << "Sending request with user's Identity: " << request << std::endl;
			if (d_identity.getSeatType() == Identity::BPS) {
				std::cout << "BPS User" << std::endl;
			} else if (d_identity.getSeatType() == Identity::NONBPS) {
				std::cout << "NON-BPS User" << std::endl;
			} else if (d_identity.getSeatType() == Identity::INVALID_SEAT) {
				std::cout << "Invalid User" << std::endl;
			}
			d_session->sendRequest(request, d_identity, CorrelationId(this));
		}

          while (true) {
              Event event = d_session->nextEvent();
              if (event.eventType() != Event::RESPONSE &&
                  event.eventType() != Event::PARTIAL_RESPONSE) {
                      continue;
              }

              MessageIterator msgIter(event);
              while (msgIter.next()) {
                  Message msg = msgIter.message();
                  if (msg.hasElement(FIELD_SEARCH_ERROR)) {
                      msg.print(std::cout);
                      continue;
                  }

                  Element categories = msg.getElement("category");
                  int numCategories = categories.numValues();

                  for (int catIdx=0; catIdx < numCategories; ++catIdx) {

                      Element category = categories.getValueAsElement(catIdx);
                      std::string Name = 
                          category.getElementAsString("categoryName");
                      std::string Id = 
                          category.getElementAsString("categoryId");

                      std::cout << "\n  Category Name:" << 
                          padString (Name, CAT_NAME_LEN) <<
                          "\tId:" << Id << std::endl;

                      Element fields = category.getElement("fieldData");
                      int numElements = fields.numValues();

                      printHeader();
                      for (int i=0; i < numElements; i++) {
                          printField (fields.getValueAsElement(i));
                      }
                  }
                  std::cout << std::endl;
              }
              if (event.eventType() == Event::RESPONSE) {
                  break;
              }
          }
      }
};

int main(int argc, char **argv){
    SimpleCategorizedFieldSearchExample example;
    try {
        example.run(argc, argv);
    } catch (Exception &e) {
        std::cerr << "Library Exception!!! " << e.description() << std::endl;
    }
    // wait for enter key to exit application
    std::cout << "Press ENTER to quit" << std::endl;
    char dummy[2];
    std::cin.getline(dummy, 2);
    return 0;
}
